frontend/src/pages/NewRunV2.tsx

```diff
@@ -120,6 +145,60 @@ function NewRunV2(props: NewRunV2Props) {
   const titleVerb = existingRunId ? 'Clone' : 'Start';
   const titleAdjective = existingRunId ? '' : 'new';
 
+  // Use pipeline, pipeline version, and experiment from parent component
+  useEffect(() => {
+    setPipeline(existingPipeline);
+    setPipelineVersion(existingPipelineVersion);
+    setApiExperiment(chosenExperiment);
+  }, [existingPipeline, existingPipelineVersion, chosenExperiment]);
+
+  // If pipeline is changed, call back to NewRunSwitcher
+  useEffect(() => {
+    if (updatedPipeline?.id) {
+      const searchString = urlParser.build({
+        [QUERY_PARAMS.experimentId]: experimentId || '',
+        [QUERY_PARAMS.pipelineId]: updatedPipeline.id || '',
+        [QUERY_PARAMS.pipelineVersionId]: '',
+      });
+      props.history.replace(searchString);
+      handlePipelineVersionIdChange('');
+      handlePipelineIdChange(updatedPipeline.id);
+    }
+  }, [updatedPipeline]);
+
+  // If pipeline version is changed, call back to NewRunSwitcher
+  useEffect(() => {
+    if (pipeline?.id && updatedPipelineVersion?.id) {
+      const searchString = urlParser.build({
+        [QUERY_PARAMS.experimentId]: experimentId || '',
+        [QUERY_PARAMS.pipelineId]: pipeline.id || '',
+        [QUERY_PARAMS.pipelineVersionId]: updatedPipelineVersion.id || '',
+      });
+      props.history.replace(searchString);
+      handlePipelineVersionIdChange(updatedPipelineVersion.id);
+    }
+  }, [updatedPipelineVersion]);
+
+  useEffect(() => {
+    if (apiExperiment?.id && pipeline?.id && pipelineVersion?.id) {
+      const searchString = urlParser.build({
+        [QUERY_PARAMS.experimentId]: apiExperiment.id || '',
+        [QUERY_PARAMS.pipelineId]: pipeline.id || '',
+        [QUERY_PARAMS.pipelineVersionId]: pipelineVersion.id || '',
+      });
+      props.history.replace(searchString);
+      return;
+    }
+
+    if (apiExperiment?.id) {
+      const searchString = urlParser.build({
+        [QUERY_PARAMS.experimentId]: apiExperiment.id || '',
+      });
+      props.history.replace(searchString);
+      return;
+    }
+  }, [apiExperiment]);
```

Comment:
Question: Can we use a callback function instead of `useEffect` for all these changes? Looks like you can simplify the logic here without using `useEffect`